src/devices/sound/nds_sound.cpp

```diff
@@ -0,0 +1,829 @@
+				break;
+			case 2: // channel 3
+				lmix = m_channel[3].loutput;
+				break;
+			case 3: // channel 1 + 3
+				lmix = m_channel[1].loutput + m_channel[3].loutput;
+				break;
+			}
+
+			switch (rout_from())
+			{
+			case 0: // right mixer
+				break;
+			case 1: // channel 1
+				rmix = m_channel[1].routput;
+				break;
+			case 2: // channel 3
+				rmix = m_channel[3].routput;
+				break;
+			case 3: // channel 1 + 3
+				rmix = m_channel[1].routput + m_channel[3].routput;
+				break;
+			}
+
+			// adjust master volume
+			lmix = (lmix * mvol()) >> 13;
+			rmix = (rmix * mvol()) >> 13;
+
+			// add bias and clip output
+			lmix = std::max<int>(0, std::min<int>(0x3ff, (lmix + (m_bias & 0x3ff))));
+			rmix = std::max<int>(0, std::min<int>(0x3ff, (rmix + (m_bias & 0x3ff))));
+
+			// set output
+			outputs[0].put_int_clamp(sampindex, lmix, 0x3ff);
+			outputs[1].put_int_clamp(sampindex, rmix, 0x3ff);
+		}
+	}
+}
+
+
+//-------------------------------------------------
+//  channel_control_r - read sound channel control
+//  register
+//-------------------------------------------------
+
+u32 nds_sound_device::channel_control_r(offs_t offset)
+{
+	return m_channel[(offset >> 2) & 0xf].control;
+}
+
+
+//-------------------------------------------------
+//  channel_w - write sound channel register
+//-------------------------------------------------
+
+void nds_sound_device::channel_w(offs_t offset, u32 data, u32 mem_mask)
+{
+	m_channel[(offset >> 2) & 0xf].write(offset & 3, data, mem_mask);
+}
+
+
+//-------------------------------------------------
+//  channel_t::write - write sound channel register
+//-------------------------------------------------
+
+void nds_sound_device::channel_t::write(offs_t offset, u32 data, u32 mem_mask)
+{
+	const u32 old = control;
+	switch (offset & 3)
+	{
+	case 0: // Control/Status
+		COMBINE_DATA(&control);
+		if (BIT(old ^ control, 31))
+			keyon();
+		else
+			keyoff();
+
+		// reset hold flag
+		if (!playing && !hold())
+		{
+			sample = lfsr_out = 0;
+			output = loutput = routput = 0;
+		}
+		break;
+	case 1: // Source address
+		mem_mask &= 0x7ffffff;
+		COMBINE_DATA(&sourceaddr);
+		break;
+	case 2: // Frequency, Loopstart
+		if (ACCESSING_BITS_0_15)
+			freq = (freq & BIT(~mem_mask, 0, 16)) | (BIT(data & mem_mask, 0, 16));
+		if (ACCESSING_BITS_16_31)
+			loopstart = (loopstart & BIT(~mem_mask, 16, 16)) | (BIT(data & mem_mask, 16, 16));
+		break;
+	case 3: // Length
+		mem_mask &= 0x3fffff;
+		COMBINE_DATA(&length);
+		break;
+	}
+}
+
+
+//-------------------------------------------------
+//  channel_t::keyon - execute sound channel keyon
+//-------------------------------------------------
+
+void nds_sound_device::channel_t::keyon()
+{
+	if (!playing)
+	{
+		playing = true;
+		delay = format() == 2 ? 11 : 3; // 3 (11 for ADPCM) delay for playing sample
+		cur_bitaddr = cur_addr = 0;
+		cur_state = format() == 2 ? STATE_ADPCM_LOAD : (loopstart == 0 ? STATE_POST_LOOP : STATE_PRE_LOOP);
```

Comment:
Ternary conditional statements are easier to read if the conditions are in parentheses, like this: `(format() == 2) ? STATE_ADPCM_LOAD : (loopstart == 0) ? STATE_POST_LOOP : STATE_PRE_LOOP;`